index.js

```diff
@@ -12,7 +12,26 @@ function isNegative(pattern) {
 	return pattern[0] === '!';
 }
 
+function isString(value) {
+	return typeof value === 'string';
+}
+
+function validatePatternsArray(patterns) {
+	if (isString(patterns)) {
+		// bail early, the rest will easier this way
+		return;
+	}
+
+	if (!Array.isArray(patterns) || !patterns.every(isString)) {
+		throw new TypeError('patterns must be a string or an array of strings');
+	}
+}
+
 function generateGlobTasks(patterns, opts) {
+	// make sure input is correct, so that we can safely parse
+	// it in this task
+	validatePatternsArray(patterns);
+
 	var globTasks = [];
 
 	patterns = arrify(patterns);
```

Comment:
That last note is exactly why I don't want to do `arrify` before validation. It will cause some invalid input to be ignored, and to me, that is weird behavior. I can do the rest a bit later today.